SwiftCompilerSources/Sources/Optimizer/Utilities/WalkUtils.swift

```diff
@@ -0,0 +1,444 @@
+//===--- WalkUtils.swift - Utilities for use-def def-use walks ------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+enum WalkResult {
+  /// stop walking further for the current use/definition
+  case stopWalk
+  /// abort the walk for all uses/definitions
+  case abortWalk
+  /// continue the walk
+  case continueWalk
+}
+
+enum WalkerVisitKind {
+  case interiorValue
+  case terminalValue
+  case unmatchedPath
+}
+
+
+protocol HasResult {
+  var result: WalkResult { get }
+  
+  func with(result: WalkResult) -> Self
+}
+
+protocol HasState {
```

Comment:
This protocol is actually not needed. `State` can be added as associated type to both, the `UseVisitor` and `DefVisitor`, protocols.